metadata-ingestion/src/datahub/ingestion/api/source_helpers.py

```diff
@@ -166,68 +169,136 @@ def auto_materialize_referenced_tags(
     stream: Iterable[MetadataWorkUnit],
+    *,
+    dry_run: bool = False,
+    drop_dirs: Sequence[str] = (),
+    platform_key: Optional[PlatformKey] = None,
 ) -> Iterable[MetadataWorkUnit]:
-    """Generate BrowsePathsV2 from Container and BrowsePaths aspects."""
+    """Generate BrowsePathsV2 from Container and BrowsePaths aspects.
 
-    ignore_urns: Set[str] = set()
-    legacy_browse_paths: Dict[str, List[str]] = defaultdict(list)
-    container_urns: Set[str] = set()
-    parent_container_map: Dict[str, str] = {}
-    children: Dict[str, List[str]] = defaultdict(list)
-    for wu in stream:
-        yield wu
+    Generates browse paths v2 on demand, rather than waiting for end of ingestion,
+    for better UI experience while ingestion is running.
 
-        urn = wu.get_urn()
-        if guess_entity_type(urn) == "container":
-            container_urns.add(urn)
-
-        container_aspects = wu.get_aspects_of_type(ContainerClass)
-        for c_aspect in container_aspects:
-            parent = c_aspect.container
-            parent_container_map[urn] = parent
-            children[parent].append(urn)
-
-        browse_path_aspects = wu.get_aspects_of_type(BrowsePathsClass)
-        for b_aspect in browse_path_aspects:
-            if b_aspect.paths:
-                path = b_aspect.paths[0]  # Only take first path
-                legacy_browse_paths[urn] = [
-                    p for p in path.strip("/").split("/") if p.strip() not in drop_dirs
+    To do this, assumes entities in container structure arrive in topological order
+    and that all relevant aspects (Container, BrowsePaths, BrowsePathsV2) for an urn
+    arrive together in a batch.
+
+    Calculates the correct BrowsePathsV2 at end of workunit stream,
+    and emits "corrections", i.e. a final BrowsePathsV2 for any urns that have changed.
+    """
+
+    # For telemetry, to see if our sources violate assumptions
+    num_out_of_order = 0
+    num_out_of_batch = 0
+
+    # Set for all containers and urns with a Container aspect
+    # Used to construct container paths while iterating through stream
+    # Assumes topological order of entities in stream
+    paths: Dict[str, List[BrowsePathEntryClass]] = {}
+
+    emitted_urns: Set[str] = set()
+    containers_used_as_parent: Set[str] = set()
+    for urn, batch in _batch_workunits_by_urn(stream):
+        container_path: Optional[List[BrowsePathEntryClass]] = None
+        legacy_path: Optional[List[BrowsePathEntryClass]] = None
+        has_browse_path_v2 = False
+
+        for wu in batch:
+            yield wu
+            if not wu.is_primary_source:
+                continue
+
+            container_aspect = wu.get_aspect_of_type(ContainerClass)
+            if container_aspect:
+                parent_urn = container_aspect.container
+                containers_used_as_parent.add(parent_urn)
+                paths[urn] = [
+                    *paths.setdefault(parent_urn, []),  # Guess parent has no parents
```

Comment:
i think it'd be more clear from a code readability perspective to split this into two lines

also, do we need paths.setdefault here or can we just use `paths.get(parent_urn, [])`. the mutation is throwing me off a bit

```
parent_path = paths.setdefault(parent_urn, [])
paths[urn] = [...]
```

Reply:
I think this used to have an impact but looking at the code, doesn't seem like it anymore. I'll just replace with the `.get(...)`